eth/handler.go

```diff
@@ -439,8 +442,21 @@ func (h *handler) BroadcastBlock(block *types.Block, propagate bool) {
 			log.Error("Propagating dangling block", "number", block.Number(), "hash", hash)
 			return
 		}
-		// Send the block to a subset of our peers
-		transfer := peers[:int(math.Sqrt(float64(len(peers))))]
+
+		var transfer []*ethPeer
+		if len(peers) <= numBroadcastBlockPeers {
+			// Send the block to all peers
+			transfer = peers
+		} else {
+			// Send the block to a subset of our peers
+			rand.Seed(time.Now().UnixNano())
+			indexes := rand.Perm(len(peers))
+			maxPeers := int(math.Sqrt(float64(len(peers))))
+			transfer = make([]*ethPeer, 0, maxPeers)
+			for i := 0; i < maxPeers; i++ {
+				transfer = append(transfer, peers[indexes[i]])
+			}
+		}
```

Comment:
Sending to all peers when there are less than 8 peers is not a good idea. With the current formula, the full block will be sent to 1 peer if there are 1-3 peers, to 2 peers if there 4-8 peers. It's sufficient to send it to the sqrt of peers because they will also propagate the block.

Shuffling the `peers` slice before selecting the peer is a good idea, and can be done using

    rand.Shuffle(len(peers), func (i, j int) { peers[i], peers[j] = peers[j], peers[i] })

Reply:
You don't need to shuffle because i's already random

Reply:
Perhaps not truly random, but from a practical perspective should be ok. The first item will be completely random and the next ones the iteration order of the map starting from the first item. Any reason that is not enough?

Reply:
Okay, but in practice, when my node is only connected to one node, the latest block td does appear to have not been updated for a long time. That may be a very small probability.

Reply:
Blocks are announced to everyone, even to those not broadcast to, so it should not happen.

Reply:
I suggested the shuffle because it's not good practice to rely on map iteration order. Go does not provide any guarantees about the order, the fact that it is random is an implementation detail. Since we already have the peers in a slice here, we might as well shuffle the slice explicitly.